book/src/engine/logic.md

```diff
@@ -0,0 +1,123 @@
+# Logic
+
+## Overview
+
+`chalk-engine` solves a `Goal` using a depth-first search. When asked to solve a
+particular `Goal` it hasn't seen before, it will first ask the [`Context`] to
+generate a set of program clauses, that get turned into `Strand`s, that could
+solve that goal. Otherwise, if asked to solve a `Goal` it has seen before, it
+will select the existing table.
+
+Once a table is selected, it will pick a `Strand` and a subgoal of that
+`Strand`, try to solve that `Goal`, repeating the process.
+
+When an `Answer` is found for a `Goal`, it is merged into the parent `Strand`,
+or returned if it was the root `Goal`. It will then go on to pick the next
+subgoal of the `Strand` and continue on.
+
+If at any point the solving stops being "successful" (i.e. we definitely found
+something to be unsolvable), the solving is restarted at the root `Goal`.
+
+## The stack
+
+In order to detect cycles (talked more about later), as well as keep track of
+the selected `Strand` for each table, `chalk-engine` stores a [`Stack`] on the
+`Forest`. Whenever a new goal is selected, a `StackEntry` is pushed onto the
+`Stack`, as well as the the "time" (which also gets incremented) that it was
+pushed. This "time" can be compared later to check if all the `Strands` of a
+`Table` have been checked in a single solve.
+
+As either `Answer`s are found for the selected `Table`, entries on the stack are
+`pop`ed. If something is found to be unsolvable, the complete stack is unwound.
+
+## Table creation
+
+As mentioned before, whenever a new `Goal` is encounted, a new `Table` is
+created to store current and future answers. First, the `Goal` is converted into
+an [`HhGoal`]. If it can be simplified, then a `Strand` with one or more
+subgoals will be generated and can be followed as above. Otherwise, if it is a
+`DomainGoal` (see above), then
+[`program_clauses`](https://rust-lang.github.io/chalk/chalk_engine/context/trait.ContextOps.html#tymethod.program_clauses)
+is called and each clause is converted into a `Strand` and can be followed.
+
+## `root_answer` and `ensure_root_answer`
+
+The `root_answer` function is the entry point to solve a `Goal`. Up until now,
+the idea of `Answer` versus `CompleteAnswer` have been ignored. However, in
+reality `Answer`s to `Goal`s may actually have delayed subgoals (see `ExClause`
+and `Coinduction and refinement strands`), whereas `CompleteAnswer`s may not.
```

Comment:
I don't think this should be "verbatim text"?
```suggestion
and [Coinduction and refinement strands], whereas `CompleteAnswer`s may not.
```

Reply:
Yeah, I wasn't sure how to link to a separate part of the doc. Changed this.